text/0000-unwinding.md

```diff
@@ -0,0 +1,116 @@
+- Start Date: 2015-01-21
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+While Rust does not have exceptions, it does do unwinding. The primary reason for unwinding is that
+it allows resources to be cleaned up in the event of an unrecoverable error. Part of Rust's model
+for failure is that failures are only catchable on task boundaries, and that the state of a failed
+task is not visible from the outside. This is a simple, clean model that allows for efficient robust
+programs.
+
+This RFC is a proposal to allow unwinding to be caught at the boundaries between non-Rust and Rust
+functions. It also provides the user with the ability specify whether a given function will unwind
+or not for cases when the function body is not available.
+
+# Motivation
+
+The major use case for this feature is to allow users to prevent and handle unwinding into non-Rust
+code. Unwinding into non-Rust code is currently considered undefined behaviour but we give users no
+tools to prevent this behaviour. While in some cases the code can be manually audited for potential
+unwinding, if arbitrary user code can be executed (for example, in a callback), then this unwinding
+is impossible to prevent.
+
+# Detailed design
+
+## Semantics
+
+The current Rust semantics are that failure leaves the task in an invalid state and therefore
+continued execution within that task is undefined behaviour. "Task" currently refers to threads,
+however this RFC proposes to extend the definition of a task to include the entry into Rust code
+from non-Rust code. As such, the task ends when control returns to original non-Rust code.
+
+This extension essentially produces "nested tasks", where the thread is one task, then each transfer
+of control from non-Rust code to Rust code is another. Unwinding out of a task is undefined
+behaviour, meaning that the current behaviour regarding unwinding into non-Rust code is preserved.
+However, the extended definition of "task" allows for unwinding to be caught at the entry point of
+the Rust code. From there a user may decide to return an error code or abort the process.
+
+Attempts to catch unwinding at any point other than a task entry point is considered undefined behaviour.
+
+## Implementation
+
+Rust already has a significant amount of code for handling unwinding. However, it is currently not
+safe to use nested invocations of the `try` function. The first part of the implementation would be
+to improve the unwinding system to allow for nested calls to the try function.
+
+The `try` function itself doesn't need changing. The current signature is `fn try<F:FnOnce()>(f: F)
+-> Result<(), Box<Any + Send>>`, which provides means for the caller to know if the invocation was
+successful. The function would remain unsafe.
```

Comment:
I think I didn't explain properly, I meant accepting the fact that if you have a function that does two writes to TLS, it may panic in the middle resulting in only a single write, and that you thus need to have a RAII struct implementing Drop if you want transactional atomicity (a simple "scoped TLS assignment" could be provided in the standard library).

It's of course also possible to have other schemes with more overhead but I think reaching any agreement on that would be quite hard.